File: packages/ts/spaces/src/registry.ts
/**
 * Space registries (plaintext metadata docs). A user's spaces live at
 * `user/{userId}/_spaces`; each space's ACCESS RECORD (owner/members + shared
 * name/image) at `spaces/{spaceId}/_access`. The object tree lives in the plaintext
 * unified object index (`objects/_index`, see `object-index.ts`); `_access` is the
 * owner-only access record. Spaces are neutral containers — visibility and encryption
 * are per-node properties (see `ObjectNode.access` / `ObjectNode.enc`).
 */
import { StarfishHttpError } from "@drakkar.software/starfish-client"
import type { StarfishClient } from "@drakkar.software/starfish-client"

import type { CapMap, PubAccessMap, Space } from "./config.js"
import type { SealedBlob } from "./config.js"
import { randomId } from "@drakkar.software/starfish-protocol"
import type { Session } from "./session.js"
import { seedSpaceObjectIndex } from "./object-index.js"
import { runCas } from "./cas-retry.js"

/** Owner-set, SHARED space identity, persisted in the `_access` registry doc
 *  (plaintext — NOT E2EE). `image` is a data URI. All fields optional for back-compat. */
export interface SpaceMeta {
  name?: string | null
  image?: string | null
}

/** A resolved name/image update fanned out so providers adopts a
 *  freshly-reconciled value without waiting for its next navigation refresh. */
export interface SpaceMetaUpdate {
  name: string
  short: string
  image?: string
}

const SPACE_SHORT_LENGTH = 2
const SPACE_FALLBACK_SUFFIX = 6

/** Build a Space object from id + name with computed `short` monogram. */
export function buildSpace(id: string, name: string, overrides?: Partial<Space>): Space {
  const trimmed = name.trim() || `space-${id.slice(-SPACE_FALLBACK_SUFFIX)}`
  return {
    id,
    name: trimmed,
    members: 1,
    ...overrides,
  }
}

const spaceMetaListeners = new Set<(spaceId: string, meta: SpaceMetaUpdate) => void>()

export function onSpaceMeta(fn: (spaceId: string, meta: SpaceMetaUpdate) => void): () => void {
  spaceMetaListeners.add(fn)
  return () => {
    spaceMetaListeners.delete(fn)
  }
}

export function broadcastSpaceMeta(spaceId: string, meta: SpaceMetaUpdate): void {
  for (const fn of spaceMetaListeners) fn(spaceId, meta)
}

export interface SpacesDoc {
  spaces: Space[]
  caps: CapMap
  pubAccess: PubAccessMap
  /** App-specific registry fields the SDK does not model. Round-tripped untouched on
   *  every CAS write so a generic-SDK mutation never drops a consumer's own data. */
  extra: Record<string, unknown>
  hash: string | null
}

/** The `_spaces` doc body sent on push — the modelled core fields plus any app-specific
 *  fields spread at the top level (never a nested `extra` key). */
type SpacesPayload = {
  spaces: Space[]
  caps: CapMap
  pubAccess: PubAccessMap
  [key: string]: unknown
}

/** Core keys the SDK models explicitly; everything else in the doc body lives in `extra`.
 *  `v` is the protocol version (re-added on write); `hash` is the CAS token (never in body). */
const CORE_SPACES_KEYS = new Set(["spaces", "caps", "pubAccess", "v", "hash"])

/** Build the `_spaces` doc body: app-specific `extra` fields spread FIRST so the modelled
 *  core fields always take precedence. No nested `extra` key reaches storage. */
function toPayload(doc: SpacesDoc): SpacesPayload {
  return {
    ...doc.extra,
    spaces: doc.spaces,
    caps: doc.caps,
    pubAccess: doc.pubAccess,
  }
}

/** Collect every doc-body key the SDK does not model into `extra`. */
function collectExtra(data: Record<string, unknown> | undefined): Record<string, unknown> {
  if (!data || typeof data !== "object") return {}
  const extra: Record<string, unknown> = {}
  for (const [k, v] of Object.entries(data)) if (!CORE_SPACES_KEYS.has(k)) extra[k] = v
  return extra
}

/** Coerce a raw `_spaces` doc body (or `undefined`) into a typed {@link SpacesDoc}. */
function coerceSpacesDoc(data: Record<string, unknown> | undefined, hash: string | null): SpacesDoc {
  const raw = data as { spaces?: Space[]; caps?: CapMap; pubAccess?: PubAccessMap } | undefined
  return {
    spaces: Array.isArray(raw?.spaces) ? raw!.spaces! : [],
    caps: raw?.caps && typeof raw.caps === "object" ? raw.caps : {},
    pubAccess: raw?.pubAccess && typeof raw.pubAccess === "object" ? raw.pubAccess : {},
    extra: collectExtra(data),
    hash,
  }
}

async function pullSpacesDoc(client: StarfishClient, session: Session): Promise<SpacesDoc> {
  const res = await client.pull(session.layout.spacesPull(session.userId)).catch((err: unknown) => {
    if (err instanceof StarfishHttpError && err.status === 404) return null
    throw err
  })
  return coerceSpacesDoc(res?.data as Record<string, unknown> | undefined, res?.hash ?? null)
}

export async function readSpaces(client: StarfishClient, session: Session): Promise<SpacesDoc> {
  try {
    return await pullSpacesDoc(client, session)
  } catch (err) {
    console.error("[readSpaces] failed to pull spaces registry", err)
    return coerceSpacesDoc(undefined, null)
  }
}

export function updateSpacesDoc(
  client: StarfishClient,
  session: Session,
  mutator: (cur: { spaces: Space[]; caps: CapMap; pubAccess: PubAccessMap }) => { spaces: Space[]; caps: CapMap; pubAccess: PubAccessMap },
): Promise<void> {
  return runCas(async () => {
    const doc = await pullSpacesDoc(client, session)
    const cur = { spaces: doc.spaces, caps: doc.caps, pubAccess: doc.pubAccess }
    const next = mutator(cur)
    if (next === cur) return
    await client.push(session.layout.spacesPush(session.userId), { v: 1 as const, ...toPayload({ ...doc, ...next }), }, doc.hash)
  })
}

/**
 * Read-modify-CAS-write ONE app-specific (`extra`) field of the `_spaces` doc.
 */
export function updateSpacesExtraField<T>(
  client: StarfishClient,
  session: Session,
  key: string,
  mutator: (cur: T | undefined) => T | null,
): Promise<void> {
  return runCas(async () => {
    const doc = await pullSpacesDoc(client, session)
    const next = mutator(doc.extra[key] as T | undefined)
    if (next === null) return
    const payload = { ...toPayload(doc), [key]: next }
    await client.push(session.layout.spacesPush(session.userId), { v: 1 as const, ...payload }, doc.hash)
  })
}

export async function writeSpaces(
  client: StarfishClient,
  session: Session,
  spaces: Space[],
): Promise<void> {
  await updateSpacesDoc(client, session, (cur) => ({ spaces, caps: cur.caps, pubAccess: cur.pubAccess }))
}

export async function reorderSpaces(client: StarfishClient, session: Session, order: string[]): Promise<void> {
  await updateSpacesDoc(client, session, (cur) => {
    const byId = new Map(cur.spaces.map((s) => [s.id, s]))
    const next: Space[] = []
    for (const id of order) {
      const s = byId.get(id)
      if (s) {
        next.push(s)
        byId.delete(id)
      }
    }
    for (const s of cur.spaces) if (byId.has(s.id)) next.push(s)
    const unchanged = next.length === cur.spaces.length && next.every((s, i) => s === cur.spaces[i])
    if (unchanged) return cur
    return { spaces: next, caps: cur.caps, pubAccess: cur.pubAccess }
  })
}

export type SpaceEntry = {
  owner: string | null
  members: string[]
  name: string | null
  image: string | null
  hash: string | null
}

export async function readSpaceAccess(
  client: StarfishClient,
  spaceId: string,
  session: Session,
): Promise<SpaceEntry> {
  const res = await client.pull(session.layout.spaceAccessPull(spaceId)).catch((err: unknown) => {
    if (err instanceof StarfishHttpError && err.status === 404) return null
    throw err
  })
  const data = res?.data as { owner?: string; members?: unknown[]; name?: string; image?: string } | undefined
  return {
    owner: typeof data?.owner === "string" ? data.owner : null,
    members: Array.isArray(data?.members)
      ? data!.members!.filter((m): m is string => typeof m === "string")
      : [],
    name: typeof data?.name === "string" ? data.name : null,
    image: typeof data?.image === "string" ? data.image : null,
    hash: res?.hash ?? null,
  }
}

export async function writeSpaceAccess(
  client: StarfishClient,
  spaceId: string,
  owner: string,
  members: string[],
  hash: string | null,
  session: Session,
  meta?: SpaceMeta,
): Promise<void> {
  const name = meta?.name?.trim() || undefined
  const image = meta?.image || undefined
  await client.push(
    session.layout.spaceAccessPush(spaceId),
    {
      v: 1, owner, members,
      ...(name ? { name } : {}),
      ...(image ? { image } : {}),
    },
    hash,
  )
}

export async function addSpaceMember(
  client: StarfishClient,
  spaceId: string,
  ownerUserId: string,
  memberUserId: string,
  session: Session,
): Promise<void> {
  const { owner, members, name, image, hash } = await readSpaceAccess(client, spaceId, session)
  if (memberUserId === (owner ?? ownerUserId) || members.includes(memberUserId)) return
  await writeSpaceAccess(client, spaceId, owner ?? ownerUserId, [...members, memberUserId], hash, session, { name, image })
}

/** Remove a member from the space roster (used for link revocation). */
export async function removeSpaceMember(
  client: StarfishClient,
  spaceId: string,
  memberUserId: string,
  session: Session,
): Promise<void> {
  const { owner, members, name, image, hash } = await readSpaceAccess(client, spaceId, session)
  if (!members.includes(memberUserId)) return
  await writeSpaceAccess(
    client,
    spaceId,
    owner ?? memberUserId,
    members.filter((m) => m !== memberUserId),
    hash,
    session,
    { name, image },
  )
}

/** Invitee/owner-side: drop a space from the identity's own list + forget its cap
 *  and link-access credential. Idempotent (no-op when absent). */
export async function removeJoinedSpace(client: StarfishClient, session: Session, spaceId: string): Promise<void> {
  await updateSpacesDoc(client, session, (cur) => {
    if (!cur.spaces.some((s) => s.id === spaceId)) return cur
    const caps = { ...cur.caps }
    delete caps[spaceId]
    const pubAccess = { ...cur.pubAccess }
    delete pubAccess[spaceId]
    return { spaces: cur.spaces.filter((s) => s.id !== spaceId), caps, pubAccess }
  })
}

/** Move one space to an absolute index in the list (clamped). */
export async function moveSpace(client: StarfishClient, session: Session, spaceId: string, toIndex: number): Promise<void> {
  await updateSpacesDoc(client, session, (cur) => {
    const from = cur.spaces.findIndex((s) => s.id === spaceId)
    if (from === -1) return cur
    const next = [...cur.spaces]
    const [moved] = next.splice(from, 1)
    next.splice(Math.max(0, Math.min(toIndex, next.length)), 0, moved)
    if (next.every((s, i) => s === cur.spaces[i])) return cur
    return { spaces: next, caps: cur.caps, pubAccess: cur.pubAccess }
  })
}

/** Append a space to the joined list (dup-guarded) plus optional cap / link-access updates. */
function addSpaceWithUpdates(
  client: StarfishClient,
  session: Session,
  space: Space,
  updates?: { caps?: CapMap; pubAccess?: PubAccessMap },
): Promise<void> {
  return updateSpacesDoc(client, session, (cur) => {
    const exists = cur.spaces.some((s) => s.id === space.id)
    if (exists && !updates) return cur
    return {
      spaces: exists ? cur.spaces : [...cur.spaces, space],
      caps: updates?.caps ? { ...cur.caps, ...updates.caps } : cur.caps,
      pubAccess: updates?.pubAccess ? { ...cur.pubAccess, ...updates.pubAccess } : cur.pubAccess,
    }
  })
}

export function addJoinedSpace(client: StarfishClient, session: Session, space: Space): Promise<void> {
  return addSpaceWithUpdates(client, session, space)
}

export function addJoinedSpaceWithCap(client: StarfishClient, session: Session, space: Space, capJson: string): Promise<void> {
  return addSpaceWithUpdates(client, session, space, { caps: { [space.id]: capJson } })
}

export function addJoinedSpaceWithLinkAccess(client: StarfishClient, session: Session, space: Space, sealed: SealedBlob): Promise<void> {
  return addSpaceWithUpdates(client, session, space, { pubAccess: { [space.id]: sealed } })
}

/**
 * Create a new space owned by the identity. Seeds an empty plaintext object index.
 */
export async function createSpace(
  session: Session,
  name: string,
): Promise<Space> {
  const { accountClient, userId } = session
  const { spaces } = await readSpaces(accountClient, session)
  const trimmed = name.trim() || "New Space"
  const id = `${session.spaceIdPrefix}${randomId()}`
  const space = buildSpace(id, trimmed)
  await writeSpaceAccess(accountClient, id, userId, [], null, session, { name: trimmed })
  await seedSpaceObjectIndex(session, id)
  await writeSpaces(accountClient, session, [...spaces, space])
  return space
}

export async function reconcileSpaceMeta(
  client: StarfishClient,
  session: Session,
  spaceId: string,
  shared: SpaceMeta,
  knownSpaces?: Space[],
): Promise<void> {
  const sharedName = typeof shared.name === "string" && shared.name.trim() ? shared.name : null
  const sharedImage = typeof shared.image === "string" && shared.image ? shared.image : null
  if (sharedName === null && sharedImage === null) return
  const known = knownSpaces?.find((s) => s.id === spaceId)
  if (known) {
    const name = sharedName ?? known.name
    const short = name.slice(0, 2).toUpperCase()
    const image = sharedImage ?? (known as Space & { image?: string }).image
    if (name === known.name && (image ?? null) === ((known as Space & { image?: string }).image ?? null)) return
    void short
  }
  const { spaces } = await readSpaces(client, session)
  const cur = spaces.find((s) => s.id === spaceId)
  if (!cur) return
  const name = sharedName ?? cur.name
  const image = sharedImage ?? (cur as Space & { image?: string }).image
  const short = name.slice(0, 2).toUpperCase()
  if (name === cur.name && (image ?? null) === ((cur as Space & { image?: string }).image ?? null)) return
  const next = spaces.map((s) => (s.id === spaceId ? { ...s, name, image } : s))
  await writeSpaces(client, session, next)
  broadcastSpaceMeta(spaceId, { name, short, image })
}

File: packages/python/spaces/starfish_spaces/registry.py
"""Space and access registries (plaintext metadata documents).

A user's spaces live at ``user/{userId}/_spaces``; each space's access record
(owner/members + shared name/image) at ``spaces/{spaceId}/_access``.  All writes
use the CAS retry helper from :mod:`starfish_spaces.cas_retry`.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Callable, Optional

from starfish_protocol.random import random_id
from starfish_sdk.types import StarfishHttpError

from starfish_spaces.cas_retry import run_cas
from starfish_spaces.object_index import seed_space_object_index

if TYPE_CHECKING:
    from starfish_sdk import StarfishClient
    from starfish_spaces.config import CapMap, PubAccessMap, SealedBlob, Space
    from starfish_spaces.session import Session


# ── Module-level meta-listener set ───────────────────────────────────────────

_space_meta_listeners: set[Callable[[str, dict[str, Any]], None]] = set()


def on_space_meta(fn: Callable[[str, dict[str, Any]], None]) -> Callable[[], None]:
    """Register a listener for ``{name, short, image?}`` meta broadcasts.

    Returns a zero-argument callable that removes the listener.
    """
    _space_meta_listeners.add(fn)
    return lambda: _space_meta_listeners.discard(fn)


def broadcast_space_meta(space_id: str, meta: dict[str, Any]) -> None:
    """Fire all registered listeners with ``space_id`` and the update dict."""
    for fn in list(_space_meta_listeners):
        try:
            fn(space_id, meta)
        except Exception:
            pass


# ── Constants ─────────────────────────────────────────────────────────────────

_CORE_SPACES_KEYS = frozenset({"spaces", "caps", "pubAccess", "v", "hash"})
_SPACE_FALLBACK_SUFFIX = 6


# ── Space builder ─────────────────────────────────────────────────────────────


def build_space(
    space_id: str,
    name: str,
    overrides: Optional[dict[str, Any]] = None,
) -> "Space":
    """Build a :class:`Space` dict with computed ``short`` monogram."""
    from starfish_spaces.config import Space  # lazy to avoid circular

    trimmed = name.strip() or f"space-{space_id[-_SPACE_FALLBACK_SUFFIX:]}"
    base: dict[str, Any] = {
        "id": space_id,
        "name": trimmed,
        "members": 1,
    }
    if overrides:
        base.update(overrides)
    return Space(**{k: base.get(k) for k in Space.__dataclass_fields__})  # type: ignore[attr-defined]


# ── SpacesDoc helpers ─────────────────────────────────────────────────────────


class SpacesDoc:
    """In-memory model of the ``_spaces`` doc (spaces list + cap/link maps + extra)."""

    __slots__ = ("spaces", "caps", "pub_access", "extra", "hash")

    def __init__(
        self,
        spaces: list[Any],
        caps: "CapMap",
        pub_access: "PubAccessMap",
        extra: dict[str, Any],
        hash: Optional[str],
    ) -> None:
        self.spaces = spaces
        self.caps = caps
        self.pub_access = pub_access
        self.extra = extra
        self.hash = hash


def _collect_extra(data: Optional[dict[str, Any]]) -> dict[str, Any]:
    if not data or not isinstance(data, dict):
        return {}
    return {k: v for k, v in data.items() if k not in _CORE_SPACES_KEYS}


def _coerce_spaces_doc(
    data: Optional[dict[str, Any]],
    hash: Optional[str],
) -> SpacesDoc:
    if not isinstance(data, dict):
        return SpacesDoc([], {}, {}, {}, hash)
    return SpacesDoc(
        spaces=list(data.get("spaces", [])) if isinstance(data.get("spaces"), list) else [],
        caps=dict(data.get("caps", {})) if isinstance(data.get("caps"), dict) else {},
        pub_access=dict(data.get("pubAccess", {})) if isinstance(data.get("pubAccess"), dict) else {},
        extra=_collect_extra(data),
        hash=hash,
    )


def _to_payload(doc: SpacesDoc) -> dict[str, Any]:
    """Spread extra FIRST so core fields always take precedence."""
    return {
        **doc.extra,
        "spaces": doc.spaces,
        "caps": doc.caps,
        "pubAccess": doc.pub_access,
    }


async def _pull_spaces_doc(client: "StarfishClient", session: "Session") -> SpacesDoc:
    try:
        res = await client.pull(session.layout.spaces_pull(session.user_id))
        data = res.data if hasattr(res, "data") else res
        hash_ = res.hash if hasattr(res, "hash") else None
        return _coerce_spaces_doc(data if isinstance(data, dict) else None, hash_)
    except StarfishHttpError as exc:
        if exc.status == 404:
            return SpacesDoc([], {}, {}, {}, None)
        raise


# ── Public readers/writers ────────────────────────────────────────────────────


async def read_spaces(client: "StarfishClient", session: "Session") -> SpacesDoc:
    """Pull the ``_spaces`` doc; returns an empty doc on any error."""
    try:
        return await _pull_spaces_doc(client, session)
    except Exception:
        return SpacesDoc([], {}, {}, {}, None)


async def update_spaces_doc(
    client: "StarfishClient",
    session: "Session",
    mutator: Callable[[dict[str, Any]], Optional[dict[str, Any]]],
) -> None:
    """CAS read-modify-write the ``_spaces`` doc.

    ``mutator`` receives ``{spaces, caps, pubAccess}`` and must return either a
    modified dict or ``None`` to skip the write.
    """
    async def attempt() -> None:
        doc = await _pull_spaces_doc(client, session)
        cur = {"spaces": doc.spaces, "caps": doc.caps, "pubAccess": doc.pub_access}
        next_ = mutator(cur)
        if next_ is None or next_ is cur:
            return
        merged = SpacesDoc(
            spaces=next_.get("spaces", doc.spaces),
            caps=next_.get("caps", doc.caps),
            pub_access=next_.get("pubAccess", doc.pub_access),
            extra=doc.extra,
            hash=doc.hash,
        )
        payload = {"v": 1, **_to_payload(merged)}
        await client.push(session.layout.spaces_push(session.user_id), payload, doc.hash)

    await run_cas(attempt)


async def update_spaces_extra_field(
    client: "StarfishClient",
    session: "Session",
    key: str,
    mutator: Callable[[Optional[Any]], Optional[Any]],
) -> None:
    """CAS update one app-specific ``extra`` field in the ``_spaces`` doc."""
    async def attempt() -> None:
        doc = await _pull_spaces_doc(client, session)
        next_val = mutator(doc.extra.get(key))
        if next_val is None:
            return
        payload = {**_to_payload(doc), key: next_val, "v": 1}
        await client.push(session.layout.spaces_push(session.user_id), payload, doc.hash)

    await run_cas(attempt)


async def write_spaces(
    client: "StarfishClient",
    session: "Session",
    spaces: list[Any],
) -> None:
    """Overwrite the ``spaces`` list (preserving caps + pubAccess)."""
    await update_spaces_doc(client, session, lambda cur: {
        "spaces": spaces, "caps": cur["caps"], "pubAccess": cur["pubAccess"]
    })


async def reorder_spaces(
    client: "StarfishClient",
    session: "Session",
    order: list[str],
) -> None:
    """Re-order the spaces list to match ``order`` (unknown ids appended)."""
    def mutator(cur: dict[str, Any]) -> Optional[dict[str, Any]]:
        by_id = {s["id"]: s for s in cur["spaces"] if isinstance(s, dict) and "id" in s}
        next_list: list[Any] = []
        seen = set()
        for id_ in order:
            if id_ in by_id and id_ not in seen:
                next_list.append(by_id[id_])
                seen.add(id_)
        # Append remaining
        for s in cur["spaces"]:
            if isinstance(s, dict) and s.get("id") not in seen:
                next_list.append(s)
        if next_list == cur["spaces"]:
            return None
        return {"spaces": next_list, "caps": cur["caps"], "pubAccess": cur["pubAccess"]}

    await update_spaces_doc(client, session, mutator)


async def move_space(
    client: "StarfishClient",
    session: "Session",
    space_id: str,
    to_index: int,
) -> None:
    """Move one space to an absolute index in the list (clamped)."""
    def mutator(cur: dict[str, Any]) -> Optional[dict[str, Any]]:
        spaces = cur["spaces"]
        from_ = next((i for i, s in enumerate(spaces) if isinstance(s, dict) and s.get("id") == space_id), -1)
        if from_ == -1:
            return None
        lst = list(spaces)
        moved = lst.pop(from_)
        idx = max(0, min(to_index, len(lst)))
        lst.insert(idx, moved)
        if lst == spaces:
            return None
        return {"spaces": lst, "caps": cur["caps"], "pubAccess": cur["pubAccess"]}

    await update_spaces_doc(client, session, mutator)


# ── SpaceEntry / _access doc ──────────────────────────────────────────────────


class SpaceEntry:
    """Model for the ``_access`` doc of a space."""

    __slots__ = ("owner", "members", "name", "image", "hash")

    def __init__(
        self,
        owner: Optional[str],
        members: list[str],
        name: Optional[str],
        image: Optional[str],
        hash: Optional[str],
    ) -> None:
        self.owner = owner
        self.members = members
        self.name = name
        self.image = image
        self.hash = hash


def _parse_space_access(data: Any, hash_: Optional[str]) -> "SpaceEntry":
    """Parse a raw ``_access`` doc body + hash into a :class:`SpaceEntry`."""
    if not isinstance(data, dict):
        return SpaceEntry(None, [], None, None, None)
    members = [m for m in data.get("members", []) if isinstance(m, str)]
    return SpaceEntry(
        owner=data.get("owner") if isinstance(data.get("owner"), str) else None,
        members=members,
        name=data.get("name") if isinstance(data.get("name"), str) else None,
        image=data.get("image") if isinstance(data.get("image"), str) else None,
        hash=hash_,
    )


async def read_space_access(
    client: "StarfishClient",
    space_id: str,
    session: "Session",
) -> SpaceEntry:
    """Pull the ``_access`` doc for ``space_id``; returns empty entry on 404."""
    try:
        res = await client.pull(session.layout.space_access_pull(space_id))
        data = res.data if hasattr(res, "data") else {}
        hash_ = res.hash if hasattr(res, "hash") else None
    except StarfishHttpError as exc:
        if exc.status == 404:
            return SpaceEntry(None, [], None, None, None)
        raise

    return _parse_space_access(data, hash_)


async def read_space_access_batch(
    session: "Session",
    space_ids: list[str],
) -> dict[str, "SpaceEntry"]:
    """Batch-read the ``_access`` doc for many spaces in a single round-trip.

    Requires the server to be configured with :func:`spaces_collections` and
    :func:`create_spaces_role_enricher` (``allow_tofu=False``, the default),
    and the caller to use the account-scoped shared client
    (``session.spaces_registry_client``) whose cap scope covers ``spaces/**``.

    Returns a ``dict[space_id, SpaceEntry]`` containing ONLY the spaces the caller
    is authorised to read.  Spaces where the server returns ``{"error": ...}``
    (e.g. caller is not a member, or the space does not exist) are silently omitted.
    """
    if not space_ids:
        return {}
    entries = await session.spaces_registry_client.batch_pull_many(
        "spaceaccess",
        [{"spaceId": sid} for sid in space_ids],
    )
    result: dict[str, SpaceEntry] = {}
    for i, space_id in enumerate(space_ids):
        if i >= len(entries):
            break
        entry = entries[i]
        if not isinstance(entry, dict) or entry.get("error"):
            continue
        result[space_id] = _parse_space_access(entry.get("data"), entry.get("hash"))
    return result


async def write_space_access(
    client: "StarfishClient",
    space_id: str,
    owner: str,
    members: list[str],
    hash: Optional[str],
    session: "Session",
    meta: Optional[dict[str, Any]] = None,
) -> None:
    """Write the ``_access`` doc for ``space_id``."""
    payload: dict[str, Any] = {"v": 1, "owner": owner, "members": members}
    name = (meta or {}).get("name", "")
    image = (meta or {}).get("image", "")
    if name and isinstance(name, str) and name.strip():
        payload["name"] = name.strip()
    if image and isinstance(image, str):
        payload["image"] = image
    await client.push(session.layout.space_access_push(space_id), payload, hash)


async def _update_space_access(
    client: "StarfishClient",
    space_id: str,
    session: "Session",
    mutator: Callable[["SpaceEntry"], Optional[dict[str, Any]]],
) -> None:
    """CAS read-modify-write the ``_access`` doc for ``space_id``.

    ``mutator`` receives the freshly-read :class:`SpaceEntry` and returns either
    ``{"owner", "members"}`` or ``None`` to skip the write.  Re-reads the current
    hash + members on every retried attempt, so a stale-hash 409 on the first
    pass is retried against the newest state (mirrors the TS ``updateSpaceAccess``).
    """
    async def attempt() -> None:
        cur = await read_space_access(client, space_id, session)
        nxt = mutator(cur)
        if nxt is None:
            return
        await write_space_access(
            client, space_id, nxt["owner"], nxt["members"], cur.hash, session,
            {"name": cur.name, "image": cur.image},
        )

    await run_cas(attempt)


async def add_space_member(
    client: "StarfishClient",
    space_id: str,
    owner_user_id: str,
    member_user_id: str,
    session: "Session",
) -> None:
    """Add ``member_user_id`` to the space roster (idempotent, CAS-retried)."""
    def mutator(cur: "SpaceEntry") -> Optional[dict[str, Any]]:
        owner = cur.owner or owner_user_id
        if member_user_id == owner or member_user_id in cur.members:
            return None
        return {"owner": owner, "members": [*cur.members, member_user_id]}

    await _update_space_access(client, space_id, session, mutator)


async def remove_space_member(
    client: "StarfishClient",
    space_id: str,
    member_user_id: str,
    session: "Session",
) -> None:
    """Remove ``member_user_id`` from the space roster (idempotent, CAS-retried)."""
    def mutator(cur: "SpaceEntry") -> Optional[dict[str, Any]]:
        if member_user_id not in cur.members:
            return None
        return {"owner": cur.owner or member_user_id,
                "members": [m for m in cur.members if m != member_user_id]}

    await _update_space_access(client, space_id, session, mutator)


# ── Registry join/remove helpers ──────────────────────────────────────────────


async def remove_joined_space(
    client: "StarfishClient",
    session: "Session",
    space_id: str,
) -> None:
    """Drop a space from the identity's own list + forget its credentials."""
    def mutator(cur: dict[str, Any]) -> Optional[dict[str, Any]]:
        if not any(s.get("id") == space_id for s in cur["spaces"] if isinstance(s, dict)):
            return None  # already absent
        caps = {k: v for k, v in cur["caps"].items() if k != space_id}
        pub_access = {k: v for k, v in cur["pubAccess"].items() if k != space_id}
        return {"spaces": [s for s in cur["spaces"] if not (isinstance(s, dict) and s.get("id") == space_id)],
                "caps": caps, "pubAccess": pub_access}

    await update_spaces_doc(client, session, mutator)


def _add_space_with_updates(
    client: "StarfishClient",
    session: "Session",
    space: Any,
    updates: Optional[dict[str, Any]] = None,
) -> Any:
    space_id = space.id if hasattr(space, "id") else space.get("id") if isinstance(space, dict) else ""
    space_dict = {"id": space_id, "name": getattr(space, "name", ""), "members": getattr(space, "members", 1)}

    def mutator(cur: dict[str, Any]) -> Optional[dict[str, Any]]:
        exists = any(s.get("id") == space_id for s in cur["spaces"] if isinstance(s, dict))
        if exists and not updates:
            return None
        caps = dict(cur["caps"])
        pub_access = dict(cur["pubAccess"])
        if updates:
            if updates.get("caps"):
                caps.update(updates["caps"])
            if updates.get("pubAccess"):
                pub_access.update(updates["pubAccess"])
        return {
            "spaces": cur["spaces"] if exists else [*cur["spaces"], space_dict],
            "caps": caps,
            "pubAccess": pub_access,
        }

    return update_spaces_doc(client, session, mutator)


def add_joined_space(client: "StarfishClient", session: "Session", space: Any) -> Any:
    """Append a space to the joined list (dup-guarded)."""
    return _add_space_with_updates(client, session, space)


def add_joined_space_with_cap(
    client: "StarfishClient", session: "Session", space: Any, cap_json: str
) -> Any:
    """Append a space to the joined list + store its cap JSON."""
    space_id = space.id if hasattr(space, "id") else space.get("id", "")
    return _add_space_with_updates(client, session, space, {"caps": {space_id: cap_json}})


def add_joined_space_with_link_access(
    client: "StarfishClient", session: "Session", space: Any, sealed: "SealedBlob"
) -> Any:
    """Append a space to the joined list + store its sealed link-access blob."""
    space_id = space.id if hasattr(space, "id") else space.get("id", "")
    sealed_dict = {"v": sealed.v, "ct": sealed.ct, "wks": sealed.wks} if hasattr(sealed, "v") else dict(sealed)
    return _add_space_with_updates(client, session, space, {"pubAccess": {space_id: sealed_dict}})


# ── Space creation ────────────────────────────────────────────────────────────


async def create_space(session: "Session", name: str) -> Any:
    """Create a new space owned by the identity, seeding its object index.

    Returns:
        The new :class:`Space` object.
    """
    client = session.account_client
    doc = await read_spaces(client, session)
    trimmed = name.strip() or "New Space"
    space_id = f"{session.space_id_prefix}{random_id()}"
    space = build_space(space_id, trimmed)
    await write_space_access(
        client, space_id, session.user_id, [], None, session, {"name": trimmed}
    )
    await seed_space_object_index(session.content_client, session, space_id)
    await write_spaces(client, session, [*doc.spaces, space])
    return space


# ── Meta reconciliation ───────────────────────────────────────────────────────


async def reconcile_space_meta(
    client: "StarfishClient",
    session: "Session",
    space_id: str,
    shared: dict[str, Any],
    known_spaces: Optional[list[Any]] = None,
) -> None:
    """Merge shared ``{name, image}`` into the local spaces list (idempotent)."""
    shared_name = shared.get("name", "")
    if isinstance(shared_name, str):
        shared_name = shared_name.strip() or None
    else:
        shared_name = None
    shared_image = shared.get("image") if isinstance(shared.get("image"), str) else None

    if shared_name is None and shared_image is None:
        return

    doc = await read_spaces(client, session)
    cur_space = next((s for s in doc.spaces if isinstance(s, dict) and s.get("id") == space_id), None)
    if not cur_space:
        return

    name = shared_name or cur_space.get("name", "")
    image = shared_image or cur_space.get("image")
    if name == cur_space.get("name") and image == cur_space.get("image"):
        return

    next_spaces = [
        ({**s, "name": name, **({"image": image} if image else {})} if s.get("id") == space_id else s)
        for s in doc.spaces
    ]
    await write_spaces(client, session, next_spaces)
    short = (name or "")[:2].upper()
    meta_update = {"name": name, "short": short}
    if image:
        meta_update["image"] = image
    broadcast_space_meta(space_id, meta_update)


__all__ = [
    "SpacesDoc",
    "SpaceEntry",
    "on_space_meta",
    "broadcast_space_meta",
    "build_space",
    "read_spaces",
    "update_spaces_doc",
    "update_spaces_extra_field",
    "write_spaces",
    "reorder_spaces",
    "move_space",
    "read_space_access",
    "write_space_access",
    "add_space_member",
    "remove_space_member",
    "remove_joined_space",
    "add_joined_space",
    "add_joined_space_with_cap",
    "add_joined_space_with_link_access",
    "create_space",
    "reconcile_space_meta",
]
